File: 4-deduced-self/message_serialiser.ts
class MessageSerialiser {
    set_source(source: string): this {
        // ...
        return this;
    }

    set_timestamp(timestamp: Date): this {
        // ...
        return this;
    }

    as_bytes(): Uint8Array {
        // ...
    }
}


class RequestSerialiser extends MessageSerialiser {
    set_request_param(param: string): this {
        // ...
        return this;
    }
}


const message = (
    new RequestSerialiser()
    .set_source('My PC')
    .set_timestamp(new Date())
    .set_request_param('give me the data!')
    .as_bytes())

File: 4-deduced-self/message_serialiser.py
from datetime import datetime
from typing import TypeVar


T = TypeVar('T', bound='MessageSerialiser')


class MessageSerialiser:
    def set_source(self: T, source: str) -> T:
        ...
        return self

    def set_timestamp(self: T, timestamp: datetime) -> T:
        ...
        return self

    def as_bytes(self) -> bytes:
        ...


class RequestSerialiser(MessageSerialiser):
    def set_request_param(self, param: str) -> 'RequestSerialiser':
        ...
        return self


message = (
    RequestSerialiser()
    .set_source('My PC')
    .set_timestamp(datetime.now())
    .set_request_param('give me the data!')
    .as_bytes())
